kvserver: add allocator sync

wenyihu6 opened the pull request:
Epic: CRDB-25222
Release note: none

----

**mmaintegration: add  allocator_op.go and allocator_sync.go**

This commit adds allocator_op.go and allocator_sync.go, introducing
trackedAllocatorChange and basic coordination helpers for allocator sync. These
helpers register changes before they are applied and post apply the effect on
store pool after the change is applied. Note that allocator sync is not plumbed
or used yet. Future commits would plumb it to replicate queue, lease queue, and
mma store rebalancer to coordinate changes.

---

**kvserver: plumb allocator sync**

Previously, allocator sync files were added. This commit wires allocator sync
into the lease queue, replicate queue, and mma store rebalancer. Each node has a
single allocator sync instance responsible for coordinating with the
corresponding mma allocator. Note that while the allocator_sync is now plumbed,
they are not yet actively used.

---

**kvserver: use allocator sync to update store pool**

Previously, allocator sync was plumbed into the replicate queue, lease queue,
and mma store rebalancer. This commit updates the lease and replicate queues to
call allocator sync to register changes before they are applied, and again post
apply to reflect load changes in the local store pool. For now, allocator sync
is only used for updating the store pool. Future commits will extend this to
coordinate with mma.

---

**kvserver: register external changes with mma**

Previously, allocator sync was called before and after applying changes, but
only to update the store pool. This commit extends it to also register changes
with the mma allocator. If mma is enabled at registration time, allocator sync
tracks the change and informs mma during PostApply. If mma is disabled, it
continues to only update the store pool.

---

**mmaintegration: add MMAPreApply to allocator sync**

Previously, allocator sync tracked external changes from the replicate and lease
queues. This commit extends it to register mma driven changes as well. On
PostApply, allocator sync now informs mma of the outcome. Note that it is not
being used yet. Future commits will call into it from mma store rebalancer.

---

**mmaintegration: integrate allocator sync to mma store rebalancer**

Previously, mma PreApply was introduced to register mma changes with allocator
sync. This commit integrates it into the mma store rebalancer. To handle
failures that occur before a change is even registered (such as when a replica
doesn’t exist), this commit also adds, MarkChangesAsFailed, in allocator sync.
This function bypasses the trackedChanges map since the change is not meant to
be applied. It must only be called on failure, and allocator sync is responsible
for notifying MMA of the failure.

---

**mmaintegration: add store load msg**

This commit adds MakeStoreLoadMsg, moved from allocator_mma_integration.go in
mmaprototypehelpers. Eventually, all functions from mmaprototypehelpers will be
migrated to pkg/kv/kvserver/mmaintegration as part of productionizing the code.
Note that no tests have been added yet - a TODO has been left in place to
address this. Added to the spreadsheet as well.

---

**mmaintegration: add InvalidSyncChangeID**

This commit adds InvalidSyncChangeID to allocator sync, used only by asim to
indicate that no change is currently in progress. This is needed because asim
returns control to the tick loop rather than waiting for changes to complete
during simulations. So it needs to track the current change id in progress.

---

**mmaintegration: remove pkg/kv/kvserver/allocator/mmaprototypehelpers**

Previously, all helper functions from
pkg/kv/kvserver/allocator/mmaprototypehelpers were migrated to
pkg/kv/kvserver/mmaintegration. This commit completes the transition by removing
all usages of mmaprototypehelpers to pkg/kv/kvserver/mmaintegration.

---

**mmaintegration: address minor code review comments**

This commit address minor code review comments.

---

**kvserver: update allocator sync on errors in AdminTransferLease**

Previously, the replicate queue could skip updating allocator sync when
rlm.AdminTransferLease failed, due to early returns on error. This commit
ensures allocator sync is always updated before exiting the function, regardless
of success or failure.

---

**mmaintegration: add interface for store pool and mma**

This commit integrates the store pool and MMA allocator into allocator sync via
an interface to facilitate testing within this package.

---

**mmaintegration: rename mmaAllocator to mmaState**

This commit cleans up a comment and renames the mmaAllocator interface to
mmaState in the mmaintegration package to avoid naming confusion
as mmaAllocator repeats the word allocator.

wenyihu6 committed:
mmaintegration: add  allocator_op.go and allocator_sync.go

This commit adds allocator_op.go and allocator_sync.go, introducing
trackedAllocatorChange and basic coordination helpers for allocator sync. These
helpers register changes before they are applied and post apply the effect on
store pool after the change is applied. Note that allocator sync is not plumbed
or used yet. Future commits would plumb it to replicate queue, lease queue, and
mma store rebalancer to coordinate changes.

wenyihu6 committed:
kvserver: register external changes with mma

Previously, allocator sync was called before and after applying changes, but
only to update the store pool. This commit extends it to also register changes
with the mma allocator. If mma is enabled at registration time, allocator sync
tracks the change and informs mma during PostApply. If mma is disabled, it
continues to only update the store pool.

wenyihu6 committed:
mmaintegration: add store load msg

This commit adds MakeStoreLoadMsg, moved from allocator_mma_integration.go in
mmaprototypehelpers. Eventually, all functions from mmaprototypehelpers will be
migrated to pkg/kv/kvserver/mmaintegration as part of productionizing the code.
Note that no tests have been added yet - a TODO has been left in place to
address this. Added to the spreadsheet as well.

wenyihu6 committed:
mmaintegration: remove pkg/kv/kvserver/allocator/mmaprototypehelpers

Previously, all helper functions from
pkg/kv/kvserver/allocator/mmaprototypehelpers were migrated to
pkg/kv/kvserver/mmaintegration. This commit completes the transition by removing
all usages of mmaprototypehelpers to pkg/kv/kvserver/mmaintegration.

wenyihu6 committed:
mmaintegration: add InvalidSyncChangeID

This commit adds InvalidSyncChangeID to allocator sync, used only by asim to
indicate that no change is currently in progress. This is needed because asim
returns control to the tick loop rather than waiting for changes to complete
during simulations. So it needs to track the current change id in progress.

wenyihu6 committed:
mmaintegration: integrate allocator sync to mma store rebalancer

Previously, mma PreApply was introduced to register mma changes with allocator
sync. This commit integrates it into the mma store rebalancer. To handle
failures that occur before a change is even registered (such as when a replica
doesn’t exist), this commit also adds, MarkChangesAsFailed, in allocator sync.
This function bypasses the trackedChanges map since the change is not meant to
be applied. It must only be called on failure, and allocator sync is responsible
for notifying MMA of the failure.

wenyihu6 committed:
kvserver: use allocator sync to update store pool

Previously, allocator sync was plumbed into the replicate queue, lease queue,
and mma store rebalancer. This commit updates the lease and replicate queues to
call allocator sync to register changes before they are applied, and again post
apply to reflect load changes in the local store pool. For now, allocator sync
is only used for updating the store pool. Future commits will extend this to
coordinate with mma.

wenyihu6 committed:
kvserver: plumb allocator sync

Previously, allocator sync files were added. This commit wires allocator sync
into the lease queue, replicate queue, and mma store rebalancer. Each node has a
single allocator sync instance responsible for coordinating with the
corresponding mma allocator. Note that while the allocator_sync is now plumbed,
they are not yet actively used.

wenyihu6 committed:
mmaintegration: add MMAPreApply to allocator sync

Previously, allocator sync tracked external changes from the replicate and lease
queues. This commit extends it to register mma driven changes as well. On
PostApply, allocator sync now informs mma of the outcome. Note that it is not
being used yet. Future commits will call into it from mma store rebalancer.

sumeerbhola reviewed (commented):
meta questions: it looks like we already have `mmaprototypehelpers.AllocatorSync` on master.
In aggregate, is this PR doing more than moving that code to a different package?
If there is more, is there a few sentence summary of the kind of changes?
If there isn't, is this about giving it a proper code review that was not done when it was added to `mmaprototypehelpers`? This PR doesn't have tests -- is this code simple enough, and will be well exercised by the asim tests? I see a lot of red in the coverage lines -- can you run the asim etc. tests with code coverage and confirm coverage?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/151369#-:-OWv8zvOEDOJKhHuyU8b:b-74o8zf)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->

wenyihu6 reviewed (commented):
The goal was mainly to land the code incrementally by breaking it up and factoring out helper functions with comments - no major changes yet. We previously merged the entire `mmaprototypehelpers` package into master without proper review, assuming it wasn’t production-facing. With this PR, I want to make sure we’re doing things more properly as we integrate with real components. I also left a few TODOs we might revisit later - they’re potentially more involved, so I held off for now.

Good point about the lack of tests. I was initially planning to add them later, but we might not get to it anytime soon, so I’ll look into it now. The asim tests use allocator sync as helpers, but they don’t really test the correctness of the package, so the coverage isn’t very meaningful. We should still ensure asim is calling it properly tho. I’ll start adding unit tests under `mmaintegration`. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/151369#-:-OWvrTxg-9eTV9Ct1wOm:blvh3xu)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->

sumeerbhola reviewed (changes requested):
> Reviewed 3 of 4 files at r3, 7 of 9 files at r4, 4 of 6 files at r6, 1 of 2 files at r8, 1 of 2 files at r9, 1 of 1 files at r10, 14 of 14 files at r11, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/151369)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained

___
*[`pkg/kv/kvserver/mma_store_rebalancer.go` line 157 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/151369#-OWwUX68An9D7Rv02Zl0:-OWwUX68An9D7Rv02Zl1:b-wppxa0) ([raw file](https://github.com/cockroachdb/cockroach/blob/bcf95c3ff27f66ab025f2d45ddf265534a756dc4/pkg/kv/kvserver/mma_store_rebalancer.go#L157)):*
> ```Go
> 	repl := m.store.GetReplicaIfExists(change.RangeID)
> 	if repl == nil {
> 		m.as.MarkChangesAsFailed(change.ChangeIDs())
> ```

Ah, now I see the use case.
___
*[`pkg/kv/kvserver/mmaintegration/mma_conversion.go` line 26 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/151369#-OWwVyAu8G5IH39_wCKk:-OWwVyAu8G5IH39_wCKl:b-dxjr6g) ([raw file](https://github.com/cockroachdb/cockroach/blob/bcf95c3ff27f66ab025f2d45ddf265534a756dc4/pkg/kv/kvserver/mmaintegration/mma_conversion.go#L26)):*
> ```Go
> ) []mmaprototype.ReplicaChange {
> 	// TODO(wenyihu6): we are passing existing replicas to
> 	// mma.RegisterExternalChanges just to get the add and remove replica state.
> ```

should this comment say `mmaprototype.MakeLeaseTransferChanges`?
___
*[`pkg/kv/kvserver/replicate_queue.go` line 1043 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/151369#-OWwUiM682yY9eiVl79Z:-OWwUiM682yY9eiVl79_:b-tb2ww6) ([raw file](https://github.com/cockroachdb/cockroach/blob/bcf95c3ff27f66ab025f2d45ddf265534a756dc4/pkg/kv/kvserver/replicate_queue.go#L1043)):*
> ```Go
> 		target,
> 	)
> 	if err := rlm.AdminTransferLease(ctx, target.StoreID, false /* bypassSafetyChecks */); err != nil {
> ```

don't we need to call PostApply with false?
___
*[`pkg/kv/kvserver/mmaintegration/allocator_sync.go` line 24 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/151369#-OWwQg_S7bM0y-JFylQF:-OWwQg_S7bM0y-JFylQG:baw98x4) ([raw file](https://github.com/cockroachdb/cockroach/blob/9837af5b00432ea9991ed597ad2fac4667141213/pkg/kv/kvserver/mmaintegration/allocator_sync.go#L24)):*
> ```Go
> // mma is disabled.)
> 
> // AllocatorSync is a component that coordinates changes from external
> ```

isn't it from "all components" with "mma and store pool"?
___
*[`pkg/kv/kvserver/mmaintegration/allocator_sync.go` line 55 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/151369#-OWwR0Zn-0DXk4bi3og0:-OWwR0Zn-0DXk4bi3og1:be52q1b) ([raw file](https://github.com/cockroachdb/cockroach/blob/9837af5b00432ea9991ed597ad2fac4667141213/pkg/kv/kvserver/mmaintegration/allocator_sync.go#L55)):*
> ```Go
> 	as.mu.Lock()
> 	defer as.mu.Unlock()
> 	as.mu.changeSeqGen += 1
> ```

nit: `as.mu.changeSeqGen++`
___
*[`pkg/kv/kvserver/mmaintegration/allocator_sync.go` line 113 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/151369#-OWwRP507cQgx730TZtJ:-OWwRP507cQgx730TZtK:b-a18ivw) ([raw file](https://github.com/cockroachdb/cockroach/blob/9837af5b00432ea9991ed597ad2fac4667141213/pkg/kv/kvserver/mmaintegration/allocator_sync.go#L113)):*
> ```Go
> 	if !success {
> 		return
> 	}
> ```

shouldn't this remove from the map, i.e., we should call `getTrackedChange` unconditionally?

----
Looks like this is fixed in a later commit.

btw, I ended up looking at all the `AllocatorSync` changes together.
___
*[`pkg/kv/kvserver/mmaintegration/allocator_sync.go` line 184 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/151369#-OWwTQRzAh6yB6geJfkg:-OWwTQS-6EANuMNyDBT7:bc2nihg) ([raw file](https://github.com/cockroachdb/cockroach/blob/010be9ef70269807a56fba384f2554c0cab073bd/pkg/kv/kvserver/mmaintegration/allocator_sync.go#L184)):*
> ```Go
> // MarkChangesAsFailed marks the given change IDs as failed without going
> // through allocator sync. This is used when mma changes fail before even
> // registering with mma via MMAPreApply.
> ```

Before even registering with `AllocatorSync` via `MMAPreApply`? I am confused by the existence of this method, but I'll read on.
___
*[`pkg/kv/kvserver/mmaintegration/allocator_sync.go` line 196 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/151369#-OWwShBA9wmftbhzXeEr:-OWwShBA9wmftbhzXeEs:b-f6ee6r) ([raw file](https://github.com/cockroachdb/cockroach/blob/010be9ef70269807a56fba384f2554c0cab073bd/pkg/kv/kvserver/mmaintegration/allocator_sync.go#L196)):*
> ```Go
> 	if changeIDs := trackedChange.changeIDs; changeIDs != nil {
> 		// Call into without checking cluster setting.
> 		as.mmaAllocator.AdjustPendingChangesDisposition(changeIDs, success)
> ```

I think your TODO at the top is dealt with here. In that, a decision is made when the change was generated on whether MMA knows about it (either generated by MMA, or registered with MMA), and that is carried through into `PostApply` regardless of a cluster setting change. Did you have some other problem case in mind?
___
*[`pkg/kv/kvserver/lease_queue.go` line 153 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/151369#-OWwUMLvDvJXn2gLoLxj:-OWwUMLvDvJXn2gLoLxk:bqzdys5) ([raw file](https://github.com/cockroachdb/cockroach/blob/bcf95c3ff27f66ab025f2d45ddf265534a756dc4/pkg/kv/kvserver/lease_queue.go#L153)):*
> ```Go
> 		lq.as.PostApply(changeID, err == nil /*success*/)
> 		if err != nil {
> 			// TODO(wenyihu6): we need to call post apply with false when as is more meaningful
> ```

I don't understand this comment, in that didn't we just call PostApply(..., false) above?
___
*[`pkg/kv/kvserver/mmaintegration/allocator_op.go` line 18 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/151369#-OWwPlegAMXz-uJb6U1W:-OWwPlegAMXz-uJb6U1X:b-dd1sk1) ([raw file](https://github.com/cockroachdb/cockroach/blob/9837af5b00432ea9991ed597ad2fac4667141213/pkg/kv/kvserver/mmaintegration/allocator_op.go#L18)):*
> ```Go
> type trackedAllocatorChange struct {
> 	usage allocator.RangeUsageInfo
> 	// Only one of the following two fields will be set.
> ```

nit: "Exactly one ..." would be more precise (I think "only one" encompasses 0, and the google search AI summary for "exactly one" vs "only one" seems to confirm that).


<!-- Sent from Reviewable.io -->

wenyihu6 committed:
mmaintegration: add interface for store pool and mma

This commit integrates the store pool and MMA allocator into allocator sync via
an interface to facilitate testing within this package.

wenyihu6 committed:
mmaintegration: address minor code review comments

This commit address minor code review comments.

wenyihu6 committed:
kvserver: update allocator sync on errors in AdminTransferLease

Previously, the replicate queue could skip updating allocator sync when
rlm.AdminTransferLease failed, due to early returns on error. This commit
ensures allocator sync is always updated before exiting the function, regardless
of success or failure.

wenyihu6 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/151369)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @sumeerbhola)

___
*[`pkg/kv/kvserver/lease_queue.go` line 153 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/151369#-OWwUMLvDvJXn2gLoLxj:-OWxfNloAf-psFpuZDwS:b-iol2ay) ([raw file](https://github.com/cockroachdb/cockroach/blob/bcf95c3ff27f66ab025f2d45ddf265534a756dc4/pkg/kv/kvserver/lease_queue.go#L153)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

I don't understand this comment, in that didn't we just call PostApply(..., false) above?
</blockquote></details>

Removed. This was a stale TODO.
___
*[`pkg/kv/kvserver/replicate_queue.go` line 1043 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/151369#-OWwUiM682yY9eiVl79Z:-OWxfUoa-iXMp5SpuAGg:bdpblmi) ([raw file](https://github.com/cockroachdb/cockroach/blob/bcf95c3ff27f66ab025f2d45ddf265534a756dc4/pkg/kv/kvserver/replicate_queue.go#L1043)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

don't we need to call PostApply with false?
</blockquote></details>

Good catch. You are right - updated. 
___
*[`pkg/kv/kvserver/mmaintegration/allocator_op.go` line 18 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/151369#-OWwPlegAMXz-uJb6U1W:-OWxfin9Cf6A_IpxaW0C:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/9837af5b00432ea9991ed597ad2fac4667141213/pkg/kv/kvserver/mmaintegration/allocator_op.go#L18)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

nit: "Exactly one ..." would be more precise (I think "only one" encompasses 0, and the google search AI summary for "exactly one" vs "only one" seems to confirm that).
</blockquote></details>

Done.
___
*[`pkg/kv/kvserver/mmaintegration/allocator_sync.go` line 24 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/151369#-OWwQg_S7bM0y-JFylQF:-OWxfkuDD9C51P_RQ1er:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/9837af5b00432ea9991ed597ad2fac4667141213/pkg/kv/kvserver/mmaintegration/allocator_sync.go#L24)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

isn't it from "all components" with "mma and store pool"?
</blockquote></details>

Done.
___
*[`pkg/kv/kvserver/mmaintegration/allocator_sync.go` line 55 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/151369#-OWwR0Zn-0DXk4bi3og0:-OWxfmx94UAuUltqbvpm:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/9837af5b00432ea9991ed597ad2fac4667141213/pkg/kv/kvserver/mmaintegration/allocator_sync.go#L55)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

nit: `as.mu.changeSeqGen++`
</blockquote></details>

Done.
___
*[`pkg/kv/kvserver/mmaintegration/allocator_sync.go` line 113 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/151369#-OWwRP507cQgx730TZtJ:-OWxgjs5B_4bNPmPsU3Z:bb6sh2i) ([raw file](https://github.com/cockroachdb/cockroach/blob/9837af5b00432ea9991ed597ad2fac4667141213/pkg/kv/kvserver/mmaintegration/allocator_sync.go#L113)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

shouldn't this remove from the map, i.e., we should call `getTrackedChange` unconditionally?

----
Looks like this is fixed in a later commit.

btw, I ended up looking at all the `AllocatorSync` changes together.
</blockquote></details>

Ack.
___
*[`pkg/kv/kvserver/mmaintegration/allocator_sync.go` line 184 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/151369#-OWwTQRzAh6yB6geJfkg:-OWxfnoF6El3TJJpeXen:b-7bkqua) ([raw file](https://github.com/cockroachdb/cockroach/blob/010be9ef70269807a56fba384f2554c0cab073bd/pkg/kv/kvserver/mmaintegration/allocator_sync.go#L184)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

Before even registering with `AllocatorSync` via `MMAPreApply`? I am confused by the existence of this method, but I'll read on.
</blockquote></details>

Noticed your other comment indicating this makes sense now. 
___
*[`pkg/kv/kvserver/mmaintegration/allocator_sync.go` line 196 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/151369#-OWwShBA9wmftbhzXeEr:-OWxg3fx2tgdWG7GQNK2:b-rjqii3) ([raw file](https://github.com/cockroachdb/cockroach/blob/010be9ef70269807a56fba384f2554c0cab073bd/pkg/kv/kvserver/mmaintegration/allocator_sync.go#L196)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

I think your TODO at the top is dealt with here. In that, a decision is made when the change was generated on whether MMA knows about it (either generated by MMA, or registered with MMA), and that is carried through into `PostApply` regardless of a cluster setting change. Did you have some other problem case in mind?
</blockquote></details>

No specific cases in mind yet. I was generally concerned about things like the replicate queue calling mma before the rebalancer receives store leaseholder msg. I don't think it's a problem, since components aren't supposed to rely on each other's internal view. Mainly a reminder to watch out for myself when working on thrashing prevention.
___
*[`pkg/kv/kvserver/mmaintegration/mma_conversion.go` line 26 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/151369#-OWwVyAu8G5IH39_wCKk:-OWxgmY5CxA_QJIdf-YK:b-bw8e14) ([raw file](https://github.com/cockroachdb/cockroach/blob/bcf95c3ff27f66ab025f2d45ddf265534a756dc4/pkg/kv/kvserver/mmaintegration/mma_conversion.go#L26)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

should this comment say `mmaprototype.MakeLeaseTransferChanges`?
</blockquote></details>

Updated.


<!-- Sent from Reviewable.io -->

sumeerbhola reviewed (changes requested):
Tests look very thorough

> Reviewed 1 of 2 files at r2, 1 of 2 files at r9, 4 of 4 files at r12, 1 of 1 files at r13, 2 of 2 files at r14, 4 of 4 files at r15, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/151369#-:-OWzq_8f1HOoxyo4oHyT:b-z24dpx)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @wenyihu6)

___
*[`pkg/kv/kvserver/mmaintegration/allocator_sync.go` line 196 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/151369#-OWwShBA9wmftbhzXeEr:-OWzfgj019yWYj4la1-6:b2nhysk) ([raw file](https://github.com/cockroachdb/cockroach/blob/010be9ef70269807a56fba384f2554c0cab073bd/pkg/kv/kvserver/mmaintegration/allocator_sync.go#L196)):*
<details><summary><i>Previously, wenyihu6 (Wenyi Hu) wrote…</i></summary><blockquote>

No specific cases in mind yet. I was generally concerned about things like the replicate queue calling mma before the rebalancer receives store leaseholder msg. I don't think it's a problem, since components aren't supposed to rely on each other's internal view. Mainly a reminder to watch out for myself when working on thrashing prevention.
</blockquote></details>

ack
___
*[`pkg/kv/kvserver/mmaintegration/mmaintegration_test.go` line 68 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/151369#-OWzjffi7PzdWOgY_xWV:-OWzjffi7PzdWOgY_xWW:b-dbt5h1) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad4e1d995b4aef105921aaba6cedbef92af922c4/pkg/kv/kvserver/mmaintegration/mmaintegration_test.go#L68)):*
> ```Go
> 	case roachpb.REMOVE_VOTER, roachpb.REMOVE_NON_VOTER:
> 		m.storeCPULoads[storeID] -= float64(usage.RequestCPUNanosPerSecond + usage.RaftCPUNanosPerSecond)
> 	}
> ```

This approach was unexpected, since the real StorePool looks at more than these 2 fields in `RangeUsageInfo`. For a real mock one would either (a) initialize it with a `*testing.T` and give it the expected values that it could assert on when the call was made, or (b) store the call parameters, and the test using the mock would have the assertions on the relevant fields of the parameters.

Alternatively, we usually do this via a datadriven test instead of assertions, where the `mockStorePool` would have a `*strings.Builder` to which it would print its parameters, and we would write out everything in the `strings.Builder` after every `TestData.Cmd` (e.g. `replica_rac2.TestProcessorBasic`).

I'll read on.
___
*[`pkg/kv/kvserver/mmaintegration/mmaintegration_test.go` line 95 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/151369#-OWzluGA9PD91EG9AiwO:-OWzluGA9PD91EG9AiwP:b-63zjzm) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad4e1d995b4aef105921aaba6cedbef92af922c4/pkg/kv/kvserver/mmaintegration/mmaintegration_test.go#L95)):*
> ```Go
> 	changeIDs := make([]mmaprototype.ChangeID, len(changes))
> 	for i := range changeIDs {
> 		id := mmaprototype.ChangeID(i + 1)
> ```

So with multiple calls to `RegisterExternalChanges` will the ids start from 1 each time? I suppose the duplication doesn't affect how this is used in the test.
___
*[`pkg/kv/kvserver/mmaintegration/mmaintegration_test.go` line 158 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/151369#-OWzmOxW5wG0LMsrhT4e:-OWzmOxW5wG0LMsrhT4f:be8vz6i) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad4e1d995b4aef105921aaba6cedbef92af922c4/pkg/kv/kvserver/mmaintegration/mmaintegration_test.go#L158)):*
> ```Go
> 	}
> 
> 	// Add the change.
> ```

I think these are better expressed as data driven test commands. It is fine if you don't want the test to also have to parse changes. It could hardcode some changes in the code in an an array of `trackedAllocatorChange`s. And the commands could refer to the change index. e.g.
`add-tracked-change index=1` and the `syncChangeID` would also get printed as the datadriven test output.
`get-tracked-change id=<int>`
`post-apply-change id=<int> success=<bool>` 
___
*[`pkg/kv/kvserver/mmaintegration/mmaintegration_test.go` line 201 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/151369#-OWznVxB4Tz0u91pAjW4:-OWznVxB4Tz0u91pAjW5:b-9r249z) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad4e1d995b4aef105921aaba6cedbef92af922c4/pkg/kv/kvserver/mmaintegration/mmaintegration_test.go#L201)):*
> ```Go
> 	}
> 
> 	transferFrom := roachpb.ReplicationTarget{StoreID: 1, NodeID: 1}
> ```

same comment here. These cases would be part of the same datadriven test.
___
*[`pkg/kv/kvserver/mmaintegration/mmaintegration_test.go` line 253 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/151369#-OWzncauAikaihYNeg8N:-OWzncauAikaihYNeg8O:b-7ropln) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad4e1d995b4aef105921aaba6cedbef92af922c4/pkg/kv/kvserver/mmaintegration/mmaintegration_test.go#L253)):*
> ```Go
> 	}
> 
> 	// Create a test change which removes a voter and adds a non-voter to store
> ```

ditto
___
*[`pkg/kv/kvserver/mmaintegration/mmaintegration_test.go` line 316 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/151369#-OWzobq39UBcxKJbpBD6:-OWzobq39UBcxKJbpBD7:b-8u7tp2) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad4e1d995b4aef105921aaba6cedbef92af922c4/pkg/kv/kvserver/mmaintegration/mmaintegration_test.go#L316)):*
> ```Go
> 		RangeID: 1,
> 		// Note: We can't easily create a ReplicaChange here since it has
> 		// unexported fields.
> ```

We should export a method `mmaprototype.MakePendingRangeChangeForTesting` that can be used to populate whatever we want. 
___
*[`pkg/kv/kvserver/mmaintegration/mmaintegration_test.go` line 334 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/151369#-OWzooBw2yTx_Ys2P8Q8:-OWzooBw2yTx_Ys2P8Q9:bez7hnf) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad4e1d995b4aef105921aaba6cedbef92af922c4/pkg/kv/kvserver/mmaintegration/mmaintegration_test.go#L334)):*
> ```Go
> 	require.False(t, ok)
> 
> 	// We can't verify the store pool easily since pendingChange is not
> ```

I don't quite understand. AllocatorSync extracts information from the `PendingRangeChange` to construct its `trackedAllocatorChange` which it then uses to call into StorePool, so if we were following the datadriven test pattern outlined earlier, we would be able to print the parameters to each call to the `testStorePool`.
___
*[`pkg/kv/kvserver/mmaintegration/mmaintegration_test.go` line 358 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/151369#-OWzpWF37F7JkoyIUrql:-OWzpWF37F7JkoyIUrqm:b-1742e5) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad4e1d995b4aef105921aaba6cedbef92af922c4/pkg/kv/kvserver/mmaintegration/mmaintegration_test.go#L358)):*
> ```Go
> 	require.Equal(t, changeFailed, mma.changes[3])
> 	require.Equal(t, changeRegistered, mma.changes[4])
> }
> ```

In summary, I suspect all the tests that call `createTestAllocatorSync` could be more simply done with a single datadriven test. And it would be easy to extend with more test cases.
___
*[`pkg/kv/kvserver/mmaintegration/mmaintegration_test.go` line 387 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/151369#-OWzpky1CpSjxdAOMcCD:-OWzpky1CpSjxdAOMcCE:b-39y4w0) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad4e1d995b4aef105921aaba6cedbef92af922c4/pkg/kv/kvserver/mmaintegration/mmaintegration_test.go#L387)):*
> ```Go
> 
> 	// Note that we can't check other fields of the ReplicaChange since they
> 	// are unexported.
> ```

This is unfortunate. We could extract the relevant "public" info using the various read methods that `AllocatorSync` uses.
___
*[`pkg/kv/kvserver/mmaintegration/mmaintegration_test.go` line 420 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/151369#-OWzqG1X1jOrHMQ2CcDD:-OWzqG1X1jOrHMQ2CcDE:b-7ropln) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad4e1d995b4aef105921aaba6cedbef92af922c4/pkg/kv/kvserver/mmaintegration/mmaintegration_test.go#L420)):*
> ```Go
> 	require.Len(t, replicaChanges, 1)
> 	// Note that we can't check other fields of the ReplicaChange since they
> 	// are unexported.
> ```

ditto
___
*[`pkg/kv/kvserver/mmaintegration/allocator_sync.go` line 36 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/151369#-OWzfFIb0CPvNtLxg3LX:-OWzfFIb0CPvNtLxg3LY:b-y4r2go) ([raw file](https://github.com/cockroachdb/cockroach/blob/b0c2731edf984587d6ba4a96985c155b9c3f92d5/pkg/kv/kvserver/mmaintegration/allocator_sync.go#L36)):*
> ```Go
> 
> // AllocatorSync is a component that coordinates changes from all components
> // (including mma/replicate/lease queue) with mma and store pool. When mma is
> ```

nit: we probably need a different term for mma-store-rebalancer and mma, since the former uses the latter and needs to tell the latter when something is enacted or fails. Otherwise it is confusing to think of mma needing to communicate with itself via AllocatorSync.
___
*[`pkg/kv/kvserver/mmaintegration/allocator_sync.go` line 39 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/151369#-OWzgek16XSgARjsjtML:-OWzgek16XSgARjsjtMM:bfkn912) ([raw file](https://github.com/cockroachdb/cockroach/blob/b8edefd6fb6f40fd069f4b1edb4a607847ec0cc9/pkg/kv/kvserver/mmaintegration/allocator_sync.go#L39)):*
> ```Go
> }
> 
> // mmaAllocator is an interface that defines the methods that the allocator sync
> ```

nit: The name mmaAllocator repeats the word allocator. How about calling this `mmaState` since these methods are primarily about updating state in mma?


<!-- Sent from Reviewable.io -->

wenyihu6 committed:
mmaintegration: rename mmaAllocator to mmaState

This commit cleans up a comment and renames the mmaAllocator interface to
mmaState in the mmaintegration package to avoid naming confusion
as mmaAllocator repeats the word allocator.

wenyihu6 reviewed (commented):
Thanks for the suggestion to explore datadriven tests - tests look cleaner and more extensible now. Put up https://github.com/cockroachdb/cockroach/pull/151510 as discussed over slack. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/151369#-:-OX1hdKU-6sNnXvDzqVJ:b-j4jsay)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @sumeerbhola)

___
*[`pkg/kv/kvserver/mmaintegration/allocator_sync.go` line 36 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/151369#-OWzfFIb0CPvNtLxg3LX:-OX1gWz622wTcfEyHJLm:bh6jpdz) ([raw file](https://github.com/cockroachdb/cockroach/blob/b0c2731edf984587d6ba4a96985c155b9c3f92d5/pkg/kv/kvserver/mmaintegration/allocator_sync.go#L36)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

nit: we probably need a different term for mma-store-rebalancer and mma, since the former uses the latter and needs to tell the latter when something is enacted or fails. Otherwise it is confusing to think of mma needing to communicate with itself via AllocatorSync.
</blockquote></details>

Renamed to mma-store-rebalancer.
___
*[`pkg/kv/kvserver/mmaintegration/allocator_sync.go` line 39 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/151369#-OWzgek16XSgARjsjtML:-OX1gf2k9pKS7epSOyJ8:b183o5k) ([raw file](https://github.com/cockroachdb/cockroach/blob/b8edefd6fb6f40fd069f4b1edb4a607847ec0cc9/pkg/kv/kvserver/mmaintegration/allocator_sync.go#L39)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

nit: The name mmaAllocator repeats the word allocator. How about calling this `mmaState` since these methods are primarily about updating state in mma?
</blockquote></details>

Renamed to mmaState. 
___
*[`pkg/kv/kvserver/mmaintegration/mmaintegration_test.go` line 68 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/151369#-OWzjffi7PzdWOgY_xWV:-OX1gg_W3-k_XSkKnIJd:b-9f7uvw) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad4e1d995b4aef105921aaba6cedbef92af922c4/pkg/kv/kvserver/mmaintegration/mmaintegration_test.go#L68)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

This approach was unexpected, since the real StorePool looks at more than these 2 fields in `RangeUsageInfo`. For a real mock one would either (a) initialize it with a `*testing.T` and give it the expected values that it could assert on when the call was made, or (b) store the call parameters, and the test using the mock would have the assertions on the relevant fields of the parameters.

Alternatively, we usually do this via a datadriven test instead of assertions, where the `mockStorePool` would have a `*strings.Builder` to which it would print its parameters, and we would write out everything in the `strings.Builder` after every `TestData.Cmd` (e.g. `replica_rac2.TestProcessorBasic`).

I'll read on.
</blockquote></details>

This will be resolved in the follow up pr. Introduced datadriven tests. 
___
*[`pkg/kv/kvserver/mmaintegration/mmaintegration_test.go` line 95 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/151369#-OWzluGA9PD91EG9AiwO:-OX1gj_h-5doQtqGr7-k:b-u4yjhe) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad4e1d995b4aef105921aaba6cedbef92af922c4/pkg/kv/kvserver/mmaintegration/mmaintegration_test.go#L95)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

So with multiple calls to `RegisterExternalChanges` will the ids start from 1 each time? I suppose the duplication doesn't affect how this is used in the test.
</blockquote></details>

Made it more realistic in the follow up pr. 
___
*[`pkg/kv/kvserver/mmaintegration/mmaintegration_test.go` line 158 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/151369#-OWzmOxW5wG0LMsrhT4e:-OX1gnpx2vJ2GTffs6HA:b-h6wdp5) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad4e1d995b4aef105921aaba6cedbef92af922c4/pkg/kv/kvserver/mmaintegration/mmaintegration_test.go#L158)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

I think these are better expressed as data driven test commands. It is fine if you don't want the test to also have to parse changes. It could hardcode some changes in the code in an an array of `trackedAllocatorChange`s. And the commands could refer to the change index. e.g.
`add-tracked-change index=1` and the `syncChangeID` would also get printed as the datadriven test output.
`get-tracked-change id=<int>`
`post-apply-change id=<int> success=<bool>`
</blockquote></details>

Addressed in the datadriven tests. I also added a parser that supports different types of replication changes to make the tests more realistic and thorough. This should also be helpful for future thrashing prevention tests.
___
*[`pkg/kv/kvserver/mmaintegration/mmaintegration_test.go` line 201 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/151369#-OWznVxB4Tz0u91pAjW4:-OX1hJWD9FwP-6rEkUul:bf89gea) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad4e1d995b4aef105921aaba6cedbef92af922c4/pkg/kv/kvserver/mmaintegration/mmaintegration_test.go#L201)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

same comment here. These cases would be part of the same datadriven test.
</blockquote></details>

Resolved in follow up pr. 
___
*[`pkg/kv/kvserver/mmaintegration/mmaintegration_test.go` line 253 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/151369#-OWzncauAikaihYNeg8N:-OX1hLl99_bE1TEPumL7:bf89gea) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad4e1d995b4aef105921aaba6cedbef92af922c4/pkg/kv/kvserver/mmaintegration/mmaintegration_test.go#L253)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

ditto
</blockquote></details>

Resolved in follow up pr. 
___
*[`pkg/kv/kvserver/mmaintegration/mmaintegration_test.go` line 316 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/151369#-OWzobq39UBcxKJbpBD6:-OX1hM3F2c9LKVk1vGZK:bf89gea) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad4e1d995b4aef105921aaba6cedbef92af922c4/pkg/kv/kvserver/mmaintegration/mmaintegration_test.go#L316)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

We should export a method `mmaprototype.MakePendingRangeChangeForTesting` that can be used to populate whatever we want.
</blockquote></details>

Resolved in follow up pr. 
___
*[`pkg/kv/kvserver/mmaintegration/mmaintegration_test.go` line 334 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/151369#-OWzooBw2yTx_Ys2P8Q8:-OX1hMc7CooFXUT9XVA4:bf89gea) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad4e1d995b4aef105921aaba6cedbef92af922c4/pkg/kv/kvserver/mmaintegration/mmaintegration_test.go#L334)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

I don't quite understand. AllocatorSync extracts information from the `PendingRangeChange` to construct its `trackedAllocatorChange` which it then uses to call into StorePool, so if we were following the datadriven test pattern outlined earlier, we would be able to print the parameters to each call to the `testStorePool`.
</blockquote></details>

Resolved in follow up pr. 
___
*[`pkg/kv/kvserver/mmaintegration/mmaintegration_test.go` line 358 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/151369#-OWzpWF37F7JkoyIUrql:-OX1hNGV3YH-F2wVN2mg:bf89gea) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad4e1d995b4aef105921aaba6cedbef92af922c4/pkg/kv/kvserver/mmaintegration/mmaintegration_test.go#L358)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

In summary, I suspect all the tests that call `createTestAllocatorSync` could be more simply done with a single datadriven test. And it would be easy to extend with more test cases.
</blockquote></details>

Resolved in follow up pr. 
___
*[`pkg/kv/kvserver/mmaintegration/mmaintegration_test.go` line 387 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/151369#-OWzpky1CpSjxdAOMcCD:-OX1hP21Cn4YbxYJAcvj:bcgcv3c) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad4e1d995b4aef105921aaba6cedbef92af922c4/pkg/kv/kvserver/mmaintegration/mmaintegration_test.go#L387)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

This is unfortunate. We could extract the relevant "public" info using the various read methods that `AllocatorSync` uses.
</blockquote></details>

The datadriven tests handle this nicely by using the `String` methods,.


<!-- Sent from Reviewable.io -->

wenyihu6 commented:
I'm not sure if dropping the last commit would mess up with reviewable, so I'm leaving it for now. I plan to drop it before merging, since your comments about exploring datadriven tests are addressed in the follow up PR, making that commit unnecessary.

sumeerbhola reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

please go ahead and merge without the test

> Reviewed 1 of 1 files at r16, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/151369#-:-OX3o3e0DWIZmJr-fiLf:b-r950g6)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @wenyihu6)



<!-- Sent from Reviewable.io -->

wenyihu6 committed:
mmaintegration: fix linter on NewAllocatorSync

wenyihu6 commented:
Last push fixes a linter failure (not sure why crlfmt missed it).
